File: Retrofit excercise/app/src/main/java/hr/ferit/davidbilic/retrofitexcercise/RecyclerAdapter.java
package hr.ferit.davidbilic.retrofitexcercise;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

public class RecyclerAdapter extends RecyclerView.Adapter<CustomViewHolder>{

    private List<Product> productList = new ArrayList<>();

    @NonNull
    @Override
    public CustomViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.recycler_cell,parent, false);
        return new CustomViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull CustomViewHolder customViewHolder, int position) {
        customViewHolder.setCellAttributes(productList.get(position));
    }

    @Override
    public int getItemCount() {
        if (productList!=null)
        {return productList.size();}
        return 0;
    }

    public void addProducts(List<Product> list){
        this.productList.clear();
        this.productList.addAll(list);
        notifyDataSetChanged();
    }
}

File: KotlinRetrofitExcercise-TvShows/app/src/main/java/gg/bidavid/tvshowsapp/RecyclerAdapter.kt
package gg.bidavid.tvshowsapp

import android.view.LayoutInflater
import android.view.SearchEvent
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.view.menu.ActionMenuItemView
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import gg.bidavid.tvshowsapp.model.SearchResult
import kotlinx.android.synthetic.main.tv_show_cell.view.*

class RecyclerAdapter(clickListener: ClickListener): RecyclerView.Adapter<RecyclerAdapter.TvShowViewHolder>() {
    val clickListener:ClickListener
    val tvShowsList :MutableList<SearchResult> = mutableListOf()

    init {
        this.clickListener=clickListener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TvShowViewHolder {
        val cellView = LayoutInflater.from(parent.context).inflate(R.layout.tv_show_cell,parent,false)
        return TvShowViewHolder(cellView,clickListener)
    }
    override fun onBindViewHolder(holder: TvShowViewHolder, position: Int) {
        holder.setPoster(tvShowsList.get(position))
    }
    override fun getItemCount(): Int = tvShowsList.size
    fun addData(newShowsList: List<SearchResult>){
        this.tvShowsList.clear()
        this.tvShowsList.addAll(newShowsList)
        notifyDataSetChanged()
    }




    class TvShowViewHolder(itemView: View, clickListener: ClickListener): RecyclerView.ViewHolder(itemView) {
        val clickListener:ClickListener

        init {
            this.clickListener=clickListener
            itemView.ivShowPoster.setOnClickListener{ clickListener.OnPosterClick(adapterPosition) }
        }

        fun setPoster(result: SearchResult){
            val url = result.show.poster?.mediumURL ?: null
            Picasso.get().load(url).placeholder(R.mipmap.ic_launcher).into(itemView.ivShowPoster)
        }

    }
}